flutter/flutter-provider/state-management-in-flutter-using-provider-Part2.md

```diff
@@ -0,0 +1,209 @@
+
+Need more clarity?
+ 
+> Sure, Other providers generally provide the single value from the `object<T>`, but `ListenableProvider` provides the whole model as the value to its dependents.
+When the state changes in Listenable, it notifies the provider about the state change. Now provider, via it's `InheritedWidget`
+marks it's listening dependent to rebuild.
+
+
+In the above example, the whole `Counter()` object is being returned from `create`, as this value is provided to the dependents, dependents use this value as `Counter()` and updates it's state and rebuilds again to show the updated value on the screen.
+
+### **ChangeNotifierProvider** 
+
+In change notifier provider everything is same as `ListenableProvider`, one additional specification of `ListenableProvider` for ChangeNotifier. It will automatically call `ChangeNotifier.dispose` when needed.
+
+ `ChangeNotifier` can be used with `ListenableProvider<T>`, but `ChangeNotifierProvider<T>` provides one additional benefit of disposing the ChangeNotifier when done using ChangeNotifier's dispose() method.
+
+### **ValueListenableProvider** 
+
+`ValueListenableProvider` gets the value change notification from  `ValueListenable`, and provides the value to its dependent as state. `ValueListenableProvider<T>` receives values of type T from a `ValueListenable<T>`.
+
+### **StreamProvider** 
+
+
+StreamProvider gets the events from a subscribed stream and provides these events(values) to its dependent widgets. we can use `StreamProvider<T>` to receive events of type T from a `Stream<T>`.
+
+```dart
+StreamProvider<int>(
+  initialData: 0,
+  builder: (context) {
+    return Stream<int>
+      .periodic(Duration(milliseconds: 100), (count) => count + 1)
+      .take(100);
+  },
+  child: ...
+)
+
+```
+In the above example, `StreamProvider` emits `count` value ranges from 0 to 100, on the interval of 100 milliseconds, dependents can use these events for some purpose.
+
+### **FutureProvider** 
+
+`FutureProvider` gets the value of a future and provides to its dependents. We can use `FutureProvider<T>` to get the `Future<T>` value and this value represents the state.
+
+```dart
+FutureProvider<int>(
+  initialData: 0,
+  builder: (context) {
+    return Future.delayed(Duration(seconds: 4), () => 1);
+  },
+  child: ...
+),
+```
+This, `FutureProvider<int>` gives the binary value 0 and 1, in the initial state it provides the 0, and as soon as future resolves (after 4 seconds)this provides the 1 as value
+
+### How to use provider in our code
+
+Provider has some methods
+
+- `context.watch<T>()`, which makes the widget listen to changes on `T`
+- `context.read<T>()`, which returns `T` without listening to it
+- `context.select<T, R>(R cb(T value))`, which allows a widget to listen to only a small part of `T`.
+
+Or we can use the `static` method `Provider.of<T>(context)` or  `Provider.of<T>(context, listen : false)` which will behave similarly to `watch`/`read` respectively
```

Comment:
Too many `or`. The reader will get confused. Keep the info you're giving to minimal. Only those are needed which explains the core Provider efficiently. Advance stuff can always come later